### ISSUE TITLE
workspace.json gets cluttered with numerous duplicate textMateRules

### ISSUE THREAD

**Issue by FaceCrap** [LuqueDaniel/vscode-language-renpy#263]
I've been trying without success to get comments displayed in italic. Doing a search on the web, I learned I had to modify the settings to include an `editor.tokenColorCustomizations` section.
But no matter what I tried, it didn't work. Through File > Preferences > Settings I ended up seeing this:

![image](https://user-images.githubusercontent.com/95778718/219851796-fafb2bba-580d-4a9c-8dce-6de5921c4edb.png)

So instead of adding it to my user `settings.json`, I switched to the Workspace tab, and clicked the `Edit in settings.json` link. 
But instead of ending up there, `workspace.json` got opened, which in my understanding would only affect the current workspace, which is not what I want.
Adding the scope and setting for comments to show in italic did work after adding it here, but only for this workspace... it just overrides anything I add in my user settings.

However, that did show me something else surprising, there were numerous duplicate scope and settings entries (of which the below screenshot is only a small sample)

![image](https://user-images.githubusercontent.com/95778718/219851727-438ab35d-87a6-4536-a325-a883a4601cdc.png)

That file is just littered with them. 

So basically this is causing me a lot of overhead, I am forced to do this for **every** workspace, plus I end up with a buttload of auto-added duplicate rules.

**Comment by duckdoom4** [comment]
Oh that's not good. I thought I had addressed this, but I guess the equals comparison operation misses some of them.

Can you confirm a couple of things for me?
- You are using the pre-release version now correct?
- Does this only happen with color keys? (#xxxxxx) or also with tag keys (italic, bold, etc.)

**Comment by FaceCrap** [comment]
Yes, I am using the pre-release version v2.1.3

Apart from this being a bug... the fact it adds this section to the workspace file also means I have to modify it for every project to have comments in italic.
So it would be nice if the extension provided a way to:

- if possible, redefine colors
- specify if a particular scope needs to be set italic, bold, etc.

so that it's automatically used for every project instead of having to manually modify that section in every workspace file

Another thing I noticed is that having the color defined in dialogue text, e.g. like `{color=#123456}`, doesn't produce the little color square to produce the color selector popup. 
Is that an extension thing ? Thought better ask first before I open that as an issue.

**Comment by duckdoom4** [comment]
So this is where visual studio themes come in. The scopes we use, are designed to be super customizable and follow the 'unwritten' standards that popular themes use.

If you wish to change your color scheme, all you have to do is install/select a different theme

The custom scopes we've added in the configs, are only there to extend the default themes. That way we can force certain styles to certain scopes like a color, italic or bold.

**Comment by duckdoom4** [comment]
And on that note, giving comments a certain style is easy that way. You use the build in scope inspector tool, to figure out the scope name of a comment and then you assign a style to this scope in your theme :)

**Comment by duckdoom4** [comment]
> Another thing I noticed is that having the color defined in dialogue text, e.g. like {color=#123456}, doesn't produce the little color square to produce the color selector popup.
> Is that an extension thing ? Thought better ask first before I open that as an issue.

As for this part, this is a know issue and a limitation of our current color detection regex patterns. We have a big update planned to a new pattern matching system, but this takes time and unfortunately we currently have little to spare. 
(Half of this system has been build, but the other half that would make it usable is still in the works)

**Comment by LuqueDaniel** [comment]
@FaceCrap We have released an update (2.1.4), can you try again?

**Comment by FaceCrap** [comment]
> We have released an update (2.1.4), can you try again?

Not sure if this is intentional, but apart from the various combinations of the bold/italic/etc. scopes and the `"scope": "renpy.meta.color.text"`, there's no color scopes at all now?
Also, in my code, where previously the text where color keys were applied, now don't show up in that color anymore.

**Comment by duckdoom4** [comment]
Hmm, that would suggest that the fix we applied broke this feature. In that case we'll need to have a look at the code again to find the cause

**Comment by LuqueDaniel** [comment]
> 'if not some'

I didn't realize it either (remainder to myself: next time I think about adding a breakpoint, I should do it 🤣).

**Comment by FaceCrap** [comment]
> I've pushed the change, 2.1.5 should now work as intended.

When will it appear in the Marketplace?

**Comment by FaceCrap** [comment]
> Hmm, that would suggest that the fix we applied broke this feature. ~In that case we'll need to have a look at the code again to find the cause.~ Ah yes, it should be 'if not some' 😅
> 
> I've pushed the change, 2.1.5 should now work as intended.

I thought adding the 👍 was enough 😊 but see the issue is still labeled as bug.
I did see an odd scope with a malformed color tag once, but after removing it, I never saw it appear again. So this may have been a left over from an improper formatting in the code that I fixed. 

Anyway, all seems to be working as intended now.
